### ISSUE TITLE
@observer component ignoring rootComponent.forceUpdate()

### ISSUE THREAD

**Issue by anri-asaturov** [mobxjs/mobx-react#148]
Excuse me if my question seems a bit vague, but there is too much code and it is too complex to demonstrate.

Long story short:
- I have a react app using latest stable versions of react, react-router and mobx.
- Some of the components are `@observer` some not. 
- Root component (the one that is rendered by react router and is always mounted   `<Route path="/" component={Root}>...`) has a `reaction` set to an observable in locale store and when it detects locale change it calls `this.forceUpdate()` to re-render all children so the ui translation strings change (`this` is the right one, not the issue).

Now the problem is that `@observer` components won't re-render. Their `shouldComponentUpdate` is not called and `render` is not called, whereas their `@observer` children might be re-rendered as expected.

Is it not possible to override mobx's `shouldComponentUpdate`?

----

TLDR: `rootComponent.forceUpdate()` is ignored by the child `@observer` components.

**Comment by mweststrate** [comment]
Afaik `forceUpdate` is not a recursive process, it only causes one component to update. 
`observer` components will rerender if they either receive new props, or observables they use change. 
Triggering a forceupdate from a reaction sounds quite strange to me. Could you express why you need that?

note that `reaction(() => expr, () => this.forceUpdate)` is effectively the same as just calling `expr()` in your render method. 

For locales it should be pretty simple, if you have a function that picks of the correct translation based on some current locale observable, and that function is called from your components, the component will rerender if you change the locale. You don't have to do anything for that, it's what MobX should do itself.

**Comment by anri-asaturov** [comment]
Eh, you are right, looks like forceUpdate only skips shouldComponentUpdate for the component it has been called on.

The whole system is about changing app's locale on-the-fly, worked perfectly before (didn't use mobx then).
```
        languageStore.loadSavedLanguage();
        this.onLanguageChange = reaction(
            () => languageStore.language,
            () => {
                this.forceUpdate();
            }
        );```

This code is in root component and then anywhere in child component you can show dropdown with languages and set `languageStore.language` to a new value. And components use {t('translationKey')} to get string in currently loaded locale. `t` is not involved in observables in any way, i wanted to keep it out bcs it's our separate translation tool to be used for many projects. But it looks like you are right and i'll have to create an observable wrapper for it or smth like that.

**Comment by andykog** [comment]
@anri-asaturov, you can implement this in a more mobx-friendly way:
```javascript
class LanguageStore = {
  @observable language = 'en';
  
  localize(key) {
    return myLocalizations[this.languge][key];
  }
}
```

then all the components that use `languageStore.localize('someKey')` will be automatically updated without need to call forceUpdate on `language` update.

**Comment by mweststrate** [comment]
ah so `t('key')` does *not* use `langaugeStore.language`? Yeah that kinda breaks the model, since your doing effectively double bookkeeping ;-). You could redefine t like: 

```
import t as originalT from "t"

export function t(key) {
  languageStore.language; // make sure mobx trackes this function as being dependent on language
  originalT(key)
}
```

**Comment by anri-asaturov** [comment]
As i said, 't' is a separate minimum-dependency library.

Thanks everyone, my main mistake was an assumption that sCU will be skipped for the entire tree, now it's all good, thanks! :-D
